# **NOMBRE: Andrés Jiménez**
# **Examen Segundo Bimestre RI**

## Instalacion de librerias y descagra de Corpus

In [1]:
!pip install -q PyMuPDF sentence-transformers chromadb gtts google-generativeai


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 7.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 102.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.5/19.5 MB 108.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 284.2/284.2 kB 26.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 9.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 91.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.6/101.6 kB 10.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.5/16.5 MB 119.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.6/65.6 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.5/72.5 kB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.4/119.4 kB 12.3 MB/s eta

In [2]:
!pip install kagglehub[pandas-datasets] rank_bm25 faiss-cpu nltk openai chromadb

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.3/31.3 MB 28.9 MB/s eta 0:00:00


In [3]:
!python -m nltk.downloader punkt stopwords

<frozen runpy>:128: RuntimeWarning: 'nltk.downloader' found in sys.modules after import of package 'nltk', but prior to execution of 'nltk.downloader'; this may result in unpredictable behaviour
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [4]:
!pip install dotenv

In [5]:
from dotenv import load_dotenv
import os

from google.colab import files
files.upload()  # Sube aquí tu kaggle.json
load_dotenv()

# Asignarlas a las variables de entorno
os.environ["KAGGLE_USERNAME"] = os.getenv("KAGGLE_USERNAME")
os.environ["KAGGLE_KEY"] = os.getenv("KAGGLE_KEY")
os.environ["GOOGLE_API_KEY"] = os.getenv("GOOGLE_API_KEY")

Saving .env to .env


In [6]:
!cat .env

KAGGLE_USERNAME=andresfj
KAGGLE_KEY=33fe34f5fff8005b808b457d279855ba
GOOGLE_API_KEY=AIzaSyBpjXV71erCj7555ylwMH4j7pJK7h3WOqs

In [7]:
!pip install opendatasets --upgrade


In [8]:
kaggle_username = os.getenv("KAGGLE_USERNAME")
kaggle_key = os.getenv("KAGGLE_KEY")

In [9]:
# Paso 1: Crear carpeta y mover kaggle.json
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

# Paso 2: Descargar el dataset de ArXiv
!kaggle datasets download -d cornell-university/arxiv

# Paso 3: Descomprimir archivo
import zipfile

with zipfile.ZipFile("arxiv.zip", 'r') as zip_ref:
    zip_ref.extractall("arxiv_dataset")


cp: cannot stat 'kaggle.json': No such file or directory
chmod: cannot access '/root/.kaggle/kaggle.json': No such file or directory
Dataset URL: https://www.kaggle.com/datasets/cornell-university/arxiv
License(s): CC0-1.0
 99% 1.46G/1.47G [00:06<00:00, 258MB/s]
100% 1.47G/1.47G [00:06<00:00, 227MB/s]


In [12]:
# Leer las primeras 5 líneas manualmente
import json

with open(file_path, 'r') as f:
    for i in range(5):
        print(json.loads(f.readline()))


{'id': '0704.0001', 'submitter': 'Pavel Nadolsky', 'authors': "C. Bal\\'azs, E. L. Berger, P. M. Nadolsky, C.-P. Yuan", 'title': 'Calculation of prompt diphoton production cross sections at Tevatron and\n  LHC energies', 'comments': '37 pages, 15 figures; published version', 'journal-ref': 'Phys.Rev.D76:013009,2007', 'doi': '10.1103/PhysRevD.76.013009', 'report-no': 'ANL-HEP-PR-07-12', 'categories': 'hep-ph', 'license': None, 'abstract': '  A fully differential calculation in perturbative quantum chromodynamics is\npresented for the production of massive photon pairs at hadron colliders. All\nnext-to-leading order perturbative contributions from quark-antiquark,\ngluon-(anti)quark, and gluon-gluon subprocesses are included, as well as\nall-orders resummation of initial-state gluon radiation valid at\nnext-to-next-to-leading logarithmic accuracy. The region of phase space is\nspecified in which the calculation is most reliable. Good agreement is\ndemonstrated with data from the Fermilab

In [13]:
import pandas as pd

file_path = "/content/arxiv_dataset/arxiv-metadata-oai-snapshot.json"

sample_lines = int(2150000 * 0.01)

# Usar chunks para cargar sólo una parte
df_iter = pd.read_json(file_path, lines=True, chunksize=sample_lines)
df_chunk = next(df_iter)

# Usar 'abstract' en lugar de 'summary'
df = df_chunk[["id", "title", "abstract", "authors", "categories"]].dropna().reset_index(drop=True)

df.head()



id                                              title  \
0  704.0001  Calculation of prompt diphoton production cros...   
1  704.0002           Sparsity-certifying Graph Decompositions   
2  704.0003  The evolution of the Earth-Moon system based o...   
3  704.0004  A determinant of Stirling cycle numbers counts...   
4  704.0005  From dyadic $\Lambda_{\alpha}$ to $\Lambda_{\a...   

                                            abstract  \
0    A fully differential calculation in perturba...   
1    We describe a new algorithm, the $(k,\ell)$-...   
2    The evolution of Earth-Moon system is descri...   
3    We show that a determinant of Stirling cycle...   
4    In this paper we show how to compute the $\L...   

                                             authors       categories  
0  C. Bal\'azs, E. L. Berger, P. M. Nadolsky, C.-...           hep-ph  
1                    Ileana Streinu and Louis Theran    math.CO cs.CG  
2                                        Hongjun Pan   physics.gen-ph  
3                                       David Callan          math.CO  
4           Wael Abu-Shammala and Alberto Torchinsky  math.CA math.FA

## 1. Preprocesamiento

In [14]:
import nltk
import string
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

# Descargar recursos necesarios
nltk.download("punkt")
nltk.download("stopwords")

# Lista de stopwords y puntuación
stop_words = set(stopwords.words("english"))
punctuation = set(string.punctuation)


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [15]:
def preprocess_text(text):
    text = text.lower()  # Minúsculas
    tokens = word_tokenize(text)  # Tokenización
    tokens = [word for word in tokens if word not in stop_words and word not in punctuation]
    return " ".join(tokens)


In [17]:
import nltk
nltk.download('punkt_tab')

df["text"] = df["title"] + " " + df["abstract"] + " " + df["authors"] + " " + df["categories"]

# Aplicar preprocesamiento
df["processed_text"] = df["text"].apply(preprocess_text)

# Mostrar los primeros resultados
df[["id", "processed_text"]].head()


[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


id                                     processed_text
0  704.0001  calculation prompt diphoton production cross s...
1  704.0002  sparsity-certifying graph decompositions descr...
2  704.0003  evolution earth-moon system based dark matter ...
3  704.0004  determinant stirling cycle numbers counts unla...
4  704.0005  dyadic \lambda_ \alpha \lambda_ \alpha paper s...

## 2. Indexacion y Embeddings


In [18]:
# Tf-IDF
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(df["processed_text"])


In [19]:
#BM25
from rank_bm25 import BM25Okapi

tokenized_corpus = [doc.split() for doc in df["processed_text"]]
bm25 = BM25Okapi(tokenized_corpus)


In [20]:
from sentence_transformers import SentenceTransformer
import faiss
import numpy as np

# Cargar modelo (pequeño y eficiente)
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")

# Generar vectores (embeddings)
embeddings = embedding_model.encode(df["processed_text"], show_progress_bar=True)

# Crear índice FAISS
faiss_index = faiss.IndexFlatL2(embeddings.shape[1])
faiss_index.add(np.array(embeddings))


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Batches:   0%|          | 0/672 [00:00<?, ?it/s]

In [25]:
print(embeddings)
df

[[-0.13622822  0.00425607  0.03150218 ... -0.01463755 -0.03306044
   0.0679653 ]
 [-0.01849809  0.0540137   0.00220346 ... -0.00284296  0.00682318
   0.03976696]
 [-0.0242293  -0.04620279  0.07549973 ... -0.04807935 -0.00598774
   0.02022023]
 ...
 [-0.03752438 -0.08226263 -0.08771987 ... -0.03897843 -0.06284337
   0.00193799]
 [-0.01034123  0.02519637 -0.00498092 ...  0.02235447  0.0029824
   0.03361325]
 [-0.08131544 -0.09500104  0.01677973 ... -0.05821996 -0.06275361
  -0.00715473]]


id                                              title  \
0      704.0001  Calculation of prompt diphoton production cros...   
1      704.0002           Sparsity-certifying Graph Decompositions   
2      704.0003  The evolution of the Earth-Moon system based o...   
3      704.0004  A determinant of Stirling cycle numbers counts...   
4      704.0005  From dyadic $\Lambda_{\alpha}$ to $\Lambda_{\a...   
...         ...                                                ...   
21495  708.3644        Inflation from IIB Superstrings with Fluxes   
21496  708.3645                                     Why Tau First?   
21497  708.3646  Complex solitons with power law behaviour in B...   
21498  708.3647  Sub-Threshold Field Emission from Thin Silicon...   
21499  708.3648  Evidence of a Broad Relativistic Iron Line fro...   

                                                abstract  \
0        A fully differential calculation in perturba...   
1        We describe a new algorithm, the $(k,\ell)$-...   
2        The evolution of Earth-Moon system is descri...   
3        We show that a determinant of Stirling cycle...   
4        In this paper we show how to compute the $\L...   
...                                                  ...   
21495    We study the conditions needed to have an ea...   
21496    Electron neutrino has been the first neutral...   
21497    Complex, localized stable solitons, characte...   
21498    We report on strongly enhanced electron mult...   
21499    We report on an analysis of XMM-Newton data ...   

                                                 authors         categories  \
0      C. Bal\'azs, E. L. Berger, P. M. Nadolsky, C.-...             hep-ph   
1                        Ileana Streinu and Louis Theran      math.CO cs.CG   
2                                            Hongjun Pan     physics.gen-ph   
3                                           David Callan            math.CO   
4               Wael Abu-Shammala and Alberto Torchinsky    math.CA math.FA   
...                                                  ...                ...   
21495                     Erandy Ramirez, Tonatiuh Matos           astro-ph   
21496  D. Fargion, D. D'Armiento, P. G. Lucentini De ...    astro-ph hep-ph   
21497  Utpal Roy, Rajneesh Atre, C. Sudheesh, C. Naga...     cond-mat.other   
21498  Hua Qin, Hyun-Seok Kim, Michael S. Westphall, ...  cond-mat.mes-hall   
21499  Sudip Bhattacharyya (NASA/GSFC, UMCP) and Tod ...           astro-ph   

                                                    text  \
0      Calculation of prompt diphoton production cros...   
1      Sparsity-certifying Graph Decompositions   We ...   
2      The evolution of the Earth-Moon system based o...   
3      A determinant of Stirling cycle numbers counts...   
4      From dyadic $\Lambda_{\alpha}$ to $\Lambda_{\a...   
...                                                  ...   
21495  Inflation from IIB Superstrings with Fluxes   ...   
21496  Why Tau First?   Electron neutrino has been th...   
21497  Complex solitons with power law behaviour in B...   
21498  Sub-Threshold Field Emission from Thin Silicon...   
21499  Evidence of a Broad Relativistic Iron Line fro...   

                                          processed_text  
0      calculation prompt diphoton production cross s...  
1      sparsity-certifying graph decompositions descr...  
2      evolution earth-moon system based dark matter ...  
3      determinant stirling cycle numbers counts unla...  
4      dyadic \lambda_ \alpha \lambda_ \alpha paper s...  
...                                                  ...  
21495  inflation iib superstrings fluxes study condit...  
21496  tau first electron neutrino first neutral lept...  
21497  complex solitons power law behaviour bose-eins...  
21498  sub-threshold field emission thin silicon memb...  
21499  evidence broad relativistic iron line neutron ...  

[21500 rows x 7 columns]

## 3. Funciones de Recuperacion

In [45]:
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
from rank_bm25 import BM25Okapi
import faiss

def normalize_embeddings(embeddings):
    norms = np.linalg.norm(embeddings, axis=1, keepdims=True)
    return embeddings / norms

def preprocess_query(query):
    return preprocess_text(query)

def search_tfidf(query, top_k=10):
    query_proc = preprocess_query(query)
    vec_query = tfidf_vectorizer.transform([query_proc])
    scores = cosine_similarity(tfidf_matrix, vec_query).ravel()
    top_indices = np.argsort(scores)[::-1][:top_k]
    results = df.iloc[top_indices][["id", "title", "abstract"]].copy()
    results["score"] = scores[top_indices]
    return results

def search_bm25(query, top_k=10):
    query_tokens = preprocess_query(query).split()
    scores = bm25.get_scores(query_tokens)
    top_indices = np.argsort(scores)[::-1][:top_k]
    results = df.iloc[top_indices][["id", "title", "abstract"]].copy()
    results["score"] = scores[top_indices]
    return results

def search_faiss(query, top_k=10):
    query_proc = preprocess_query(query)
    query_vec = embedding_model.encode([query_proc])
    query_vec_norm = query_vec / np.linalg.norm(query_vec, axis=1, keepdims=True)
    query_vec_norm = query_vec_norm.astype('float32')
    D, I = faiss_index_cosine.search(query_vec_norm, top_k)
    results = df.iloc[I[0]][["id", "title", "abstract"]].copy()
    results["score"] = D[0]  # Add the scores to the results DataFrame
    return results

In [46]:
queries = [
    "diphoton production cross sections",
    "quantum chromodynamics",
    "higgs boson decay",
    "machine learning for particle physics",
    "top quark production"
]

# Crear el archivo queries.txt
with open("queries.txt", "w") as f:
    for q in queries:
        f.write(q + "\n")


In [50]:
def print_results(queries_file="queries.txt", top_k=10):
    with open(queries_file, "r") as f:
        queries = [line.strip() for line in f.readlines() if line.strip()]

    for query in queries:
        print(f"\n{'='*100}\nQuery: {query}\n{'='*100}")

        for method, search_func in [
            ("TF-IDF", search_tfidf),
            ("BM25", search_bm25),
            ("FAISS", search_faiss)
        ]:
            print(f"\n--- {method} Results ---")
            results = search_func(query, top_k=top_k)
            for _, row in results.iterrows():
                snippet = row["abstract"][:300].replace("\n", " ")
                print(f"Score: {row['score']:.4f} | ID: {row['id']}\nTitle: {row['title']}\nAbstract: {snippet}...\n")


In [51]:
print_results("queries.txt", top_k=3)



Query: diphoton production cross sections

--- TF-IDF Results ---
Score: 0.3877 | ID: 704.0001
Title: Calculation of prompt diphoton production cross sections at Tevatron and
  LHC energies
Abstract:   A fully differential calculation in perturbative quantum chromodynamics is presented for the production of massive photon pairs at hadron colliders. All next-to-leading order perturbative contributions from quark-antiquark, gluon-(anti)quark, and gluon-gluon subprocesses are included, as well as a...

Score: 0.3563 | ID: 707.2294
Title: Search for a High-Mass Diphoton State and Limits on Randall-Sundrum
  Gravitons at CDF
Abstract:   We have performed a search for new particles which decay to two photons using 1.2/fb of integrated luminosity from p-pbar collisions at sqrt(s) = 1.96 TeV collected using the CDF II Detector at the Fermilab Tevatron. We find the diphoton mass spectrum to be in agreement with the standard model exp...

Score: 0.3072 | ID: 706.0851
Title: Electroweak measurem

## 4. RAG

In [52]:
def generar_contexto_faiss(query, top_k=3):
    # Codificar el embedding de la consulta
    query_proc = preprocess_query(query)
    query_vec = embedding_model.encode([query_proc])
    query_vec_norm = query_vec / np.linalg.norm(query_vec, axis=1, keepdims=True)
    query_vec_norm = query_vec_norm.astype('float32')

    # Buscar en el índice FAISS
    D, I = faiss_index_cosine.search(query_vec_norm, top_k)

    # Extraer los documentos relevantes
    textos_topk = []
    for i in I[0]:
        row = df.iloc[i]
        texto = f"Title: {row['title']}\nAbstract: {row['abstract']}"
        textos_topk.append(texto)

    # Concatenar con separadores
    contexto = "\n\n-----\n\n".join(textos_topk)

    # Mostrar el contexto preparado
    print("Contexto para el LLM:")
    print(contexto[:1500] + "\n...") if len(contexto) > 1500 else print(contexto)

    return contexto


In [54]:
query_text = "higgs boson decay"  # Ejemplo para la query
contexto = generar_contexto_faiss(query_text, top_k=3)

print(contexto)

Contexto para el LLM:
Title: Invisibly decaying Higgs boson in the Littlest Higgs model with T-parity
Abstract:   We show that there are regions in the parameter space of the Littlest Higgs
model with T-parity, allowed by electroweak precision data, where the Higgs
boson can decay invisibly into a pair of heavy photons A_H with a substantial
branching ratio. For a symmetry breaking scale f in the range 450-600 GeV, the
BR(H -> A_H A_H) can be up to 95% for an intermediate mass Higgs, and from 20%
down to a few percents for a Higgs boson of mass 200 GeV or above. The total
decay width of the Higgs boson can thereby be enhanced by an order of magnitude
compared to the Standard Model for Higgs masses around 130 GeV.


-----

Title: Effect of Charged Scalar Loops on Photonic Decays of a Fermiophobic
  Higgs
Abstract:   Higgs bosons with very suppressed couplings to fermions ("Fermiophobic Higgs
bosons", h_f) can decay to two photons (\gamma\gamma) with a branching ratio
significantly large

In [30]:
!pip install --upgrade openai


In [65]:
from IPython.display import display, Markdown
from openai import OpenAI
import os

# Assuming 'contexto' and 'query_text' are defined in previous cells

# Initialize the OpenAI client (assuming API key is set in environment variables or defined)
# If you are using a different setup for the API key, adjust this part
# client = OpenAI(api_key=os.getenv("OPENAI_API_KEY")) # Example using environment variable

# If you are using OpenRouter with a specific base_url and key:
client = OpenAI(api_key="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", base_url="https://openrouter.ai/api/v1")


modelo = "deepseek/deepseek-r1:free"

prompt=(
    f"You are an expert assistant.\n\n"
    f"Use only the provided context to answer the question.\n"
    f"If the answer is not in the context, respond with: 'No contengo esa información.'\n\n"
    f"If mathematical formulas appear, explain them clearly and write them using LaTeX syntax between $$...$$ and $...$.\n\n"
    f"If the answer is in English, translate it entirely into Spanish.\n\n"
    f"Your task:\n"
    f"1. Provide a clear and concise answer to the user's question.\n"
    f"2. Summarize the most relevant information from the context.\n"
    f"3. Justify why the retrieved documents are relevant to the question.\n\n"
    f"Context:\n\n{contexto}\n\n"
    f"Question: {query_text}"
)



response = client.chat.completions.create(
    model=modelo,
    messages=[
        {"role": "system", "content": "You answer questions based only on provided context."},
        {"role": "user", "content": prompt}
    ],
    temperature=0
)

display(Markdown(response.choices[0].message.content))

**Respuesta:**  
El bosón de Higgs puede decaer de varias formas según el modelo teórico considerado:  

1. **Modelo Littlest Higgs con T-paridad:**  
   - El Higgs puede decaer invisiblemente en dos fotones pesados ($A_H A_H$) con una razón de ramificación (BR) de hasta 95% para masas intermedias del Higgs, especialmente cuando la escala de ruptura de simetría ($f$) está entre 450-600 GeV. Esto aumenta el ancho total de decaimiento del Higgs en un orden de magnitud respecto al Modelo Estándar.  

2. **Modelo de Doble Higgs (Tipo I) con Higgs fermiofóbico:**  
   - Un Higgs con acoplamientos débiles a fermiones ($h_f$) puede decaer en dos fotones ($\gamma\gamma$) con una BR significativamente mayor que en el Modelo Estándar para masas $m_{h_f} < 150$ GeV. Bucles de Higgs cargados ($H^+$) contribuyen de forma constructiva o destructiva, afectando la precisión en mediciones futuras.  

3. **Modelos supersimétricos con neutralinos ligeros:**  
   - En el MSSM, neutralinos ($\chi$) con masas menores a 50 GeV pueden ser un canal de decaimiento dominante del Higgs ($h \to \chi\chi$), suprimiendo modos visibles como $h \to \gamma\gamma$.  

**Justificación de relevancia:**  
- **Primer documento:** Explica decaimientos invisibles del Higgs en fotones pesados, relevante para parámetros fuera del Modelo Estándar.  
- **Segundo documento:** Describe cómo los bucles de Higgs cargados modifican el decaimiento a fotones, clave para Higgs fermiofóbicos.  
- **Tercer documento:** Analiza la supresión de canales visibles del Higgs debido a neutralinos, importante en supersimetría.  

Todos los contextos abordan mecanismos de decaimiento del Higgs en escenarios teóricos alternativos, destacando modificaciones en las razones de ramificación y efectos en colisionadores.

## 5. Evaluacion

In [70]:
def evaluar_queries(archivo="queries.txt", top_k=10):
    with open(archivo, "r") as f:
        queries = [line.strip() for line in f if line.strip()]

    for query in queries:
        print(f"\n{'='*120}\n🧪 Evaluación para la consulta: '{query}'\n{'='*120}")

        # Obtener resultados
        r_tfidf = search_tfidf(query, top_k)
        r_bm25 = search_bm25(query, top_k)
        r_faiss = search_faiss(query, top_k)

        # IDs
        ids_tfidf = list(r_tfidf["id"])
        ids_bm25 = list(r_bm25["id"])
        ids_faiss = list(r_faiss["id"])

        set_tfidf, set_bm25, set_faiss = set(ids_tfidf), set(ids_bm25), set(ids_faiss)

        # Comparación de intersecciones
        comunes_tf_bm = set_tfidf & set_bm25
        comunes_tf_fa = set_tfidf & set_faiss
        comunes_bm_fa = set_bm25 & set_faiss
        comunes_todos = set_tfidf & set_bm25 & set_faiss

        print("🔁 Documentos en común:")
        print(f"TF-IDF ∩ BM25: {len(comunes_tf_bm)} documentos")
        print(f"TF-IDF ∩ FAISS: {len(comunes_tf_fa)} documentos")
        print(f"BM25 ∩ FAISS: {len(comunes_bm_fa)} documentos")
        print(f"TF-IDF ∩ BM25 ∩ FAISS: {len(comunes_todos)} documentos")

        # Comparar orden
        print("\n📊 Ordenamiento de IDs (top 5):")
        print("TF-IDF:", ids_tfidf[:5])
        print("BM25  :", ids_bm25[:5])
        print("FAISS :", ids_faiss[:5])

        # Medir similitud entre rankings
        def contar_coincidencias(lista1, lista2):
            return sum(1 for doc in lista1 if doc in lista2)

        print("\n🔢 Similitud entre rankings (top-10):")
        print(f"TF-IDF vs BM25 : {contar_coincidencias(ids_tfidf, ids_bm25)} coincidencias")
        print(f"TF-IDF vs FAISS: {contar_coincidencias(ids_tfidf, ids_faiss)} coincidencias")
        print(f"BM25 vs FAISS  : {contar_coincidencias(ids_bm25, ids_faiss)} coincidencias")

        # Tabla comparativa
        print("\n📋 Tabla comparativa (top-5):")
        print("{:<6} {:<15} {:<15} {:<15}".format("Rank", "TF-IDF", "BM25", "FAISS"))
        for i in range(5):
            print("{:<6} {:<15} {:<15} {:<15}".format(
                i+1,
                ids_tfidf[i] if i < len(ids_tfidf) else "-",
                ids_bm25[i] if i < len(ids_bm25) else "-",
                ids_faiss[i] if i < len(ids_faiss) else "-"
            ))


In [72]:
evaluar_queries("queries.txt", top_k=5)



🧪 Evaluación para la consulta: 'diphoton production cross sections'
🔁 Documentos en común:
TF-IDF ∩ BM25: 2 documentos
TF-IDF ∩ FAISS: 3 documentos
BM25 ∩ FAISS: 1 documentos
TF-IDF ∩ BM25 ∩ FAISS: 1 documentos

📊 Ordenamiento de IDs (top 5):
TF-IDF: [704.0001, 707.2294, 706.0851, 708.1277, 705.4313]
BM25  : [704.0001, 706.2117, 706.2813, 705.4313, 708.1443]
FAISS : [708.1277, 704.0001, 705.2744, 706.0851, 707.2375]

🔢 Similitud entre rankings (top-10):
TF-IDF vs BM25 : 2 coincidencias
TF-IDF vs FAISS: 3 coincidencias
BM25 vs FAISS  : 1 coincidencias

📋 Tabla comparativa (top-5):
Rank   TF-IDF          BM25            FAISS          
1      704.0001        704.0001        708.1277       
2      707.2294        706.2117        704.0001       
3      706.0851        706.2813        705.2744       
4      708.1277        705.4313        706.0851       
5      705.4313        708.1443        707.2375       

🧪 Evaluación para la consulta: 'quantum chromodynamics'
🔁 Documentos en común:
TF

#¿Cuáles documentos aparecen en común?
  Como se observa en la ejecucion del codigo, los documentos que aparecen en comun para la consulta "diphoton production cross sections": son el 707.1591 que aparece en los tres métodos de búsqueda, o el 704.0001 que aparece tanto para TF-IDF y BM25 en primer lugar y para FAISS en segundo lugar dando a entender que la similitud de busqueda y recuperacion de documentos es bastante similar entre ambos modelos de recuperación y de igual forma guarda estrecha relevancia con el modelo de FAISS, de igual forma para las demas consultas realizadas.
#¿Qué diferencias hay en el ordenamiento?
  El ordenamiento varia minimamente, debido principalmente a la forma de busqueda que se realiza, mientras TF-IDF busca por similitud coseno, BM25 lo realiza por ocurrencias de terminos, sin embargo, dentro del TOP 5, la mayoría de documentos recuperados si guarda relación.


#  Diferencias entre modelos y utilidad del RAG.
  Cuando se utilizan modelos como TF-IDF y BM25, se busca rapidez y exactitud al buscar coincidencias, ademas FAISS, tambien ofrece mejores resultados semanticos. Uniendo esto a la utilidad del RAG, ofrece una ayuda extra ya que con el ragm las respuestas son generadas de forma justificada y sin alucinar, dando como resultado una recuperacion precisa y respuestas mas coherentes y justificadas
